tfx_addons/huggingface_pusher/README.md

```diff
@@ -0,0 +1,117 @@
+
+Supported models hosted on the HuggingFace Model Hub can be directly loaded/used with APIs provided by [transformers](https://huggingface.co/docs/transformers/index) package. However, it is not limited. We can host arbitrary types of models too. 
+
+HuggingFace Space Hub provides free resources to host prototype applications that use machine learning models. It supports both private and public modes. Currently supported application frameworks are Gradio and Streamlit. It is often a good idea to host the current version of the model to the HuggingFace Space, so it can be tested before the production deployment. But you can host your models HuggingFace Spaces for other purposes as well.
+
+By keeping these information in mind, `HFModelPusher` let you push a trained or blessed model to the HuggingFace Model Hub within a new branch within TFX pipeline. Then, if specified, it pushes an application to the HuggingFace Space Hub by injecting the current model information into the prepared template sources.
+
+## Project Implementation
+HFModelPusher is a class-based TFX component, and it inherits from TFX standard `Pusher` component.
+
+### Behaviour
+
+1. Creates HuggingFace Hub __Model__ repository. It will clone one if there is already existing repository
+2. Checks out a new branch with the name of current model version. Since the model is pushed for experimental purpose, it would be good to track the versions of the model within separate branches (When the model is ready to be open to public, one can manually merge the right version(branch) into the main branch)
+3. Download all the model related files(i.e. `SavedModel`) generated from the upstream component such as `Trainer` into the model repository directory. The model files could be stored in GCS bucket, so `tf.io.gfile` module is a good choice since it handles files in location agnostic manner (GCS or local)
+4. Add & commit the current status of the Model repository
+5. Pushes the commit to the remote HuggingFace Model repository
+
+6. Creates HuggingFace Hub __Space__ repository. It will clone one if there is already an existing repository
+7. Copy all the application related files(in `app_path`) into the space repository directory 
+8. Replace special tokens in every files under the repository directory.
+    - there are two special tokens of $HF_MODEL_REPO_NAME and $HF_MODEL_REPO_BRANCH. Each special tokens(placeholders) will be replaced by the id and branch of the model repository in the step 5. 
+9. Add & commit the current status of the Space repository
+10. Pushes the commit to the __main__ branch of the HuggingFace Space Repository
+    - only the app on the main branch can be built and run, so the branch for the Space repository should not be managed
+
+
+### Structure
+It takes the following inputs:
+```
+HFModelPusher(
+    username: str,
+    hf_access_token: str,
```

Comment:
is this a private token? If so we may want to insert it in a more protected way...

Reply:
Can you guide me how to? I was injecting this token via GitHub Secret in Action

Reply:
or is it enough to add more descriptive information how to pass this token? like GitHub Action.

Reply:
There are 2 ways to go about this: 
- Encrypt your token and have a way to decrypt it (see @hanneshapke work -> https://github.com/tensorflow/tfx-addons/tree/main/tfx_addons/message_exit_handler#usage-example-at-digits )
- Using something like GCP SecretManager where you store the Secret and retrieve its value at execution time -> https://cloud.google.com/vertex-ai/docs/pipelines/secret-manager

Reply:
thank you, so I will write as how `message_exit_handler` achive this goal by optionally letting users to specify `decrypt_fn`.